ath79: add support for Dongwon T&I DW02-412H

rapid-renard opened the pull request:
Dongwon T&I DW02-412H is a 2.4/5GHz band 11ac (WiFi-5) router, based on Qualcomm Atheros QCA9557.

Specifications
--------------

- SoC: Qualcomm Atheros QCA9557-AT4A
- RAM: DDR2 128MB
- Flash: SPI NOR 2MB (Winbond W25Q16DVSSIG / ESMT F25L16PA) + NAND 64/128MB
- WiFi:
  - 2.4GHz: QCA9557 WMAC
  - 5GHz: QCA9882-BR4A
- Ethernet: 5x 10/100/1000Mbps
  - Switch: QCA8337N-AL3C
- USB: 1x USB 2.0
- UART:
  - JP2: 3.3V, TX, RX, GND (3.3V is the square pad) / 115200 8N1

Installation
--------------

1.  Connect a serial interface to UART header and interrupt the autostart of kernel.
2.  Transfer the factory image via TFTP and write it to the NAND flash.
3.  Update U-Boot environment variable.
```
> tftpboot 0x81000000 <your image>-factory.img
> nand erase 0x1000000
> nand write 0x81000000 0x1000000 ${filesize}
> setenv bootpart 2
> saveenv
```

Revert to stock firmware
--------------

1.  Revert to stock U-Boot environment variable.
```
> setenv bootpart 1
> saveenv
```

MAC addresses as verified by OEM firmware
--------------

```
   WAN: *:XX (label)
   LAN: *:XX + 1
  2.4G: *:XX + 3
    5G: *:XX + 4
```

The label MAC address was found in art 0x0.

Credits
--------------

Credit goes to the @manatails who first developed how to port OpenWRT to this device and had a significant impact on this patch.

And thanks to @adschm and @mans0n for guiding me to revise the code in many ways.

Signed-off-by: Jihoon Han rapid_renard@renard.ga

adschm commented on target/linux/ath79/dts/qca9557_kt_dw02-412h-64m.dts:
```diff
@@ -0,0 +1,38 @@
+
+		firmware@1000000 {
+			label = "firmware";
+			reg = <0x1000000 0x7000000>;
```
firmware _and_ kernel _and_ ubi? You should decide for one approach.

Apart from that, size of the firmware partition is wrong here.

mans0n replied:
Please try multi-level layout like this:
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7621_netis_wf2881.dts#L71

adschm commented on target/linux/ath79/dts/qca9557_kt_dw02-412h-64m.dts:
```diff
@@ -0,0 +1,38 @@
+
+		partition@0 {
+			label = "current";
+			reg = <0x0 0x1000000>;
```
read-only?

rapid-renard replied:
It's where the stock backup firmware is inside.
I'll change it read-only.

adschm commented on target/linux/ath79/dts/qca9557_kt_dw02-412h.dtsi:
```diff
@@ -0,0 +1,191 @@
+
+/ {
+	model = "KT DW02-412H (64M)";
+	compatible = "kt,dw02-412h-64m", "qca,qca9557";
```
Drop these two lines for obvious reasons.

rapid-renard replied:
I basically edited this file from .dts file, and I made a mistake with saving the modifications on this file before pushing it.
Sorry for making you check these again...

adschm commented on target/linux/ath79/dts/qca9557_kt_dw02-412h.dtsi:
```diff
@@ -0,0 +1,191 @@
+
+			partition@1d0000 {
+				label = "nvram";
+				reg = <0x1d0000 0x010000>;
```
read-only for these two as well?

rapid-renard replied:
I found out It is only used in another Realtek SoC based version of this device.
I'll change it read-only.

adschm commented on target/linux/ath79/dts/qca9557_kt_dw02-412h.dtsi:
```diff
@@ -0,0 +1,191 @@
+				reg = <0x050000 0x010000>;
+			};
+
+			partition@1d0000 {
```
What's between 0x60000 and 0x1d0000?

rapid-renard replied:
`mtdparts=ath-nor0:256k(boot),64k(benv),64k(log),896k(recoverk),576k(recoverr),128K(free),64k(nvram),64k(nvbackup),64k(art)`

It appears to be recovery partitions. I'll add these partitions and mark read-only.

adschm commented on target/linux/ath79/dts/qca9557_kt_dw02-412h.dtsi:
```diff
@@ -0,0 +1,191 @@
+			reg = <0x2000000 0x2000000>;
+		};
+	};
+};
```
Why _again_?

rapid-renard replied:
I basically edited this file from .dts file, and I made a mistake with saving the modifications on this file before pushing it.
Sorry for making you check these again...

adschm commented on target/linux/ath79/image/nand.mk:
```diff
@@ -150,6 +154,40 @@ define Device/glinet_gl-e750
+  DEVICE_MODEL := DW02-412H
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-storage kmod-ath10k-ct \
+	ath10k-firmware-qca988x-ct
+  SUPPORTED_DEVICES += dw02-412h
```
drop this.

adschm commented on target/linux/ath79/image/nand.mk:
```diff
@@ -150,6 +154,40 @@ define Device/glinet_gl-e750
+  KERNEL_INITRAMFS := kernel-bin | append-dtb | lzma | uImage lzma | kt-headers
+  UBINIZE_OPTS := -E 5
+  IMAGES += factory.img
+   IMAGE/factory.img := append-kernel | pad-to $$$$(KERNEL_SIZE) | append-ubi | \
```
wrong indent.

adschm commented on target/linux/ath79/image/nand.mk:
```diff
@@ -150,6 +154,40 @@ define Device/glinet_gl-e750
+
+define Device/kt_dw02-412h-64m
+  $(Device/kt_dw02-412h)
+  DEVICE_VARIANT := (64m)
```
(64M)

adschm commented on target/linux/ath79/nand/base-files/etc/board.d/01_leds:
```diff
@@ -11,6 +11,10 @@ glinet,gl-ar300m-nand|\
 	;;
+kt,dw02-412h-64m|\
+kt,dw02-412h-128m)
+	ucidef_set_led_netdev "wan" "WAN" "green:lan" "eth0.2"
```
?

rapid-renard replied:
Okay. Then... should I use this trigger? :
`ucidef_set_led_switch`

adschm replied:
yes.

adschm commented on target/linux/ath79/nand/base-files/etc/hotplug.d/ieee80211/10-fix-wifi-mac:
```diff
@@ -15,6 +15,12 @@ case $board in
+	kt,dw02-412h-128m)
+		# Set mac address for 2.4g device
+		[ "$PHYNBR" -eq 1 ] && \
+			macaddr_add $(mtd_get_mac_binary art 0x0) 3 > /sys${DEVPATH}/macaddress
```
use mtd-mac-address in DTS instead?

adschm commented on target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -22,6 +22,11 @@ case "$FIRMWARE" in
+	kt,dw02-412h-64m|\
+	kt,dw02-412h-128m)
+		caldata_extract "art" 0x5000 0x844
+		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary art 0x0) +4)
```
Nothing useful in 0x5006?

adschm commented on target/linux/ath79/dts/qca9557_kt_dw02-412h.dtsi:
```diff
@@ -0,0 +1,191 @@
+
+&usb0 {
+	#address-cells = <1>;
+	#size-cells = <0>;
```
These two can be dropped without a child node.

adschm commented on target/linux/ath79/dts/qca9557_kt_dw02-412h.dtsi:
```diff
@@ -0,0 +1,191 @@
+
+&pcie0 {
+	status = "okay";
+};
```
What that used for?

rapid-renard replied:
I basically edited this file from .dts file, and I made a mistake with saving the modifications on this file before pushing it.
Sorry for making you check these again...

rapid-renard replied:
I found out `ath10k_pci` is on `&pcie0,` not `&pcie1.`
I'll remove `&pcie1` block and migrate its content to `&pcie0`.

adschm commented on target/linux/ath79/image/gen_kt_img.sh:
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
```
Not sure whether this should go somewhere else, e.g. into /scripts ...

adschm replied:
And btw: do you need bash?

rapid-renard replied:
My develeopment environment is Ubuntu, so when I call using `#!/bin/sh,` it calls dash and gets error :
`./gen_kt_img.sh: 8: Syntax error: "(" unexpected`
so I changed it to bash.

rapid-renard replied:
> Not sure whether this should go somewhere else, e.g. into /scripts ...

In a similar case, for UniElec U7623, script gen_mt7623_emmc_img.sh was located in the target/linux/mediatek/image.
So I put this script this folder...

adschm replied:
> > And btw: do you need bash?
> 
> My develeopment environment is Ubuntu, so when I call using `#!/bin/sh`,
> it calls `dash` and gets error :
> `./gen_kt_img.sh: 8: Syntax error: "(" unexpected`
> so I changed it to `bash`.

You have to remove the "function" keyword.

rapid-renard replied:
> > > And btw: do you need bash?
> > 
> > 
> > My develeopment environment is Ubuntu, so when I call using `#!/bin/sh`,
> > it calls `dash` and gets error :
> > `./gen_kt_img.sh: 8: Syntax error: "(" unexpected`
> > so I changed it to `bash`.
> 
> You have to remove the "function" keyword.

Okay, now it works with `#!/bin/sh`. Thanks!!

adschm commented on target/linux/ath79/image/gen_kt_img.sh:
```diff
@@ -0,0 +1,52 @@
+  
+  dd bs=64 skip=1 if=${1} of=${1}$SUFFIX oflag=append conv=notrunc
+  
+  yes | cp -f ${1}$SUFFIX ${1}
```
Isn't the point of -f that you don't need input?

Apart from that, shellcheck tells me the pipe isn't working anyway.

adschm commented on target/linux/ath79/image/gen_kt_img.sh:
```diff
@@ -0,0 +1,52 @@
+  rm -f ${TMP_FILE1}
+  rm -f ${TMP_FILE2}
+  rm -f ${TMP_FILE3}
+  rm -f ${1}$SUFFIX
```
This can be put into a single line.

adschm commented on target/linux/ath79/image/gen_kt_img.sh:
```diff
@@ -0,0 +1,52 @@
+
+  dd if=${1} of=${TMP_FILE1} count=64 bs=1
+
+  dd if=${TMP_FILE1} of=${TMP_FILE2} count=4 bs=1
```
Why not read directly from $1?

adschm commented on target/linux/ath79/image/gen_kt_img.sh:
```diff
@@ -0,0 +1,52 @@
+    exit 1
+  fi
+
+  dd if=${1} of=${TMP_FILE1} count=64 bs=1
```
I don't get why you create so many intermediate files. This should all be possible like this:

```
dd if=$1 of=$1.tmp count=1 bs=4
printf '\x00\x00\x00\x00\x00\x02\x04\x00' | dd of=$1.tmp bs=1 seek=4 conv=notrunc
dd bs=1 skip=8 if=$1 >> $1.tmp

CRC32_CHECKSUM=$(crc32 $1.tmp)
echo "*** CRC32 Checksum: ${CRC32_CHECKSUM}"

echo "${CRC32_CHECKSUM}" | xxd -r -p | dd of=$1.tmp bs=1 seek=4 conv=notrunc

dd bs=64 skip=1 if=$1 of=$1.tmp oflag=append conv=notrunc

mv $1.tmp $1
```

There is still room for improvement here, i.e. dd vs. >> etc.

mans0n commented on target/linux/ath79/dts/qca9557_kt_dw02-412h.dtsi:
```diff
@@ -0,0 +1,191 @@
+	aliases {
+		led-boot = &led_wan;
+		led-failsafe = &led_wan;
+		led-running = &led_wan;
```
If you want to use this LED for WAN, you should drop this line.

rapid-renard replied:
There are no other LEDs left for these functions.
So I'm concerned that at least the failsafe status cannot be shown by LED.
Is there any alternative method to keep this function work?

mans0n replied:
I think we can use the LED for boot/failsafe/upgrade purpose (as they are special situations) but we cannot use it for WAN and running at the same time (though in practice running will be overridden by WAN). By dropping *this* line I meant only `led-running`.

rapid-renard replied:
Ah, I misundertood it as meaning to delete all these lines, sorry for my bad English...

mans0n commented on target/linux/ath79/dts/qca9557_kt_dw02-412h.dtsi:
```diff
@@ -0,0 +1,191 @@
+			label = "WPS button";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+			debounce-interval = <60>;
```
IMHO you can drop this line unless required.

mans0n commented on target/linux/ath79/dts/qca9557_kt_dw02-412h.dtsi:
```diff
@@ -0,0 +1,191 @@
+		};
+
+		wifi5g {
+			label = "green:wifi5g";
```
On stock FW this LED is used for both 2g and 5g, so strictly "wifi5g" name is wrong IMO.

rapid-renard replied:
I couldn't find a way to use this one LED for both 2g and 5g. Do you know how to make this work?

adschm replied:
Well, the name should be what it's used for with the vendor firmware. You should not just pick one, somebody else would probably decide otherwise.
If this is linked to more than one wifi in the OEM firmware, we cannot reproduce that. Best choice is to not set it at all and leave it for the user to decide.

rapid-renard replied:
That would be the right way, thanks!

mans0n commented on target/linux/ath79/dts/qca9557_kt_dw02-412h.dtsi:
```diff
@@ -0,0 +1,191 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
```
Probably you can increase the frequency up to 50MHz.

mans0n commented on target/linux/ath79/dts/qca9557_kt_dw02-412h.dtsi:
```diff
@@ -0,0 +1,191 @@
+			};
+
+			partition@1e0000 {
+				label = "nvbackup";
```
I don't think allocating two partitions for the nvram would be useful for OpenWrt users, how about consolidating "nvram" and "nvbackup" into a single partition? (Just a suggestion.)

rapid-renard replied:
I think leaving these partitions' information as in the stock firmware would be helpful for future, such as analyzing them later.

mans0n replied:
Okay, fair enough.

mans0n reviewed (commented):
Hi, a few comments below:

mans0n commented:
IMO the flash support commit should come before the device support commit.
Have you submitted this patch to upstream?

I'm not sure which one DEVICE_VENDOR should be, KT (ISP) or Dongwon (manufacturer).
It seems in most cases the actual manufacturer name is used for it ([1](https://openwrt.org/toh/alpha/asl26555) [2](https://github.com/openwrt/openwrt/commit/9a1f4b2f83ef581edca1de6f07e0d84816610f6a)). I think I had found some exceptions somewhere before but I can't remember it...

mans0n commented on target/linux/generic/pending-5.4/483-mtd-spi-nor-add-esmt-f25l16pa.patch:
```diff
@@ -0,0 +1,10 @@
+--- a/drivers/mtd/spi-nor/spi-nor.c
```
Have you submitted this patch to upstream?

rapid-renard replied:
I didn't... Should I summit this patch separately?

mans0n replied:
No, not really, though you can send it to the upstream kernel as it's a rather simple patch. I've confused pending with backport, so I think it is fine at this state, please forget this.

mans0n commented on target/linux/ath79/image/gen_kt_img.sh:
```diff
@@ -0,0 +1,52 @@
+  printf '\x00\x00\x00\x00\x00\x02\x04\x00' | dd of=${TMP_FILE2} bs=1 seek=4 conv=notrunc
+  dd bs=1 skip=8 if=${TMP_FILE1} >> ${TMP_FILE2}
+
+  CRC32_CHECKSUM=$(crc32 ${TMP_FILE2})
```
You should replace crc32 to somewhat generic utility (or include the util itself in the patch) as we cannot be sure crc32 command will exist in other platforms.
This might be a good reference: https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/image/mt7621.mk#L51

rapid-renard replied:
Thank you! I was able to add a header modifying funciton inside the Makefile thanks to the reference you provided!

rapid-renard commented:
> I'm not sure which one DEVICE_VENDOR should be, KT (ISP) or Dongwon (manufacturer).
> It seems in most cases the actual manufacturer name is used for it ([1](https://openwrt.org/toh/alpha/asl26555) [2](https://github.com/openwrt/openwrt/commit/9a1f4b2f83ef581edca1de6f07e0d84816610f6a)). I think I had found some exceptions somewhere before but I can't remember now...

The manufacturer's name of this device has been changed too often as you described in [this article](https://mans0n.github.io/2020/08/18/korean-home-router-survey), and this device is manufactured exclusively for KT.
So I think it would be good to leaving its vendor name as KT.

mans0n commented:
I also think that "KT" is shorter ;) and much eaiser to find. But there are dozens of manufacturers that supply devices to KT, and I'm not sure grouping all these devices under a single name "KT" is a good practice.

Maybe this could be a good compromise, adding DEVICE_ALT0_* here.
https://github.com/openwrt/openwrt/blob/master/target/linux/lantiq/image/danube.mk#L18

rapid-renard commented:
I think your suggestion is the proper way. Thank you for your kind reply!

mans0n commented on target/linux/ath79/dts/qca9557_dongwon_dw02-412h-64m.dts:
```diff
@@ -0,0 +1,43 @@
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
```
FYI you can reduce duplicate code further by adding labels for partitions.
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7628an_jotale_js76x8.dtsi#L89
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7628an_jotale_js76x8-16m.dts#L10
But that will make the code more hard to read, so feel free to ignore this.

adschm replied:
I think in this case partitions are few enough that having them twice should be bearable.

mans0n commented on target/linux/ath79/dts/qca9557_dongwon_dw02-412h-64m.dts:
```diff
@@ -0,0 +1,43 @@
+
+			partition@0 {
+				label = "kernel";
+				reg = <0x0 0x1000000>;
```
I don't think we need that much space for kernel... How about shrinking this to 8MB? (though that will make the numbers slightly dirtier)

rapid-renard replied:
I also thought that space was too large for kernel. I'll reduce it to 8MB.

mans0n commented on target/linux/ath79/dts/qca9557_dongwon_dw02-412h.dtsi:
```diff
@@ -0,0 +1,166 @@
+			label = "Reset button";
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+			debounce-interval = <60>;
```
Please drop this line as well.

mans0n commented on target/linux/ath79/dts/qca9557_dongwon_dw02-412h.dtsi:
```diff
@@ -0,0 +1,166 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			uboot: partition@0 {
```
Please remove unused labels.

mans0n commented on target/linux/ath79/image/nand.mk:
```diff
@@ -1,4 +1,17 @@
 # attention: only zlib compression is allowed for the boot fs
+define Build/kt-headers
```
I could not find any other device using this header format, so I think `kt-headers` is too inclusive. How about including the vendor or the device name in the function name?

And you should add this function above the zlib comment line.

rapid-renard replied:
I named it because the images of devices from other manufacturers (e.g., Mercury) also use this type of header.
However, since there is little chance that other devices will be supported, it seems good to change its name.

mans0n replied:
> I named it because the images of devices from other manufacturers (e.g., Mercury) also use this type of header.

Could you name a few? (for cross-check)

rapid-renard replied:
> Could you name a few? (for cross-check)

"GiGA WiFi home" is manufactured by two manufacturers, "Dongwon T&I" and "Mercury".
There are five models: DW01-410H, DW02-412H, KM05-410H, KM06-506H, and KM06-704H.
They all use the same firmware made by KT.

mans0n replied:
Are you sure they share the same FW? KM06-506H is based on a Realtek CPU and its firmware image header does not seem to deviate from what is commonly used on other Realtek devices.

rapid-renard replied:
> Are you sure they share the same FW? KM06-506H is based on a Realtek CPU and its firmware image header does not seem to deviate from what is commonly used on other Realtek devices.

I found a kernel module used for realtek chips in DW02-412H's firmware, so I thought all of these devices used the same firmware. I think I might have made a hasty conclusion...

mans0n commented on target/linux/ath79/image/nand.mk:
```diff
@@ -1,4 +1,17 @@
+define Build/kt-headers
+	dd if=$@ of=$@.tmp bs=4 count=1 && \
+	printf '\x00\x00\x00\x00\x00\x02\x04\x00' | dd of=$@.tmp bs=1 seek=4 conv=notrunc && \
+	dd bs=8 skip=1 if=$@ of=$@.tmp oflag=append conv=notrunc && \
```
`bs=8` here will slow down the copy speed a lot. The easiest solution is `skip_bytes`, but I think we should avoid using it on host (as macOS dd does not support the option) so I hope there would be a better way...

rapid-renard replied:
I found found a way to replace this, but I don't know if this method will be faster than the previous one.
`tail -c +9 $@ >> @.tmp`

rapid-renard replied:
This new method seems to reduce processing time very much.

`# time tail -c +9 1 >> 1.tmp`
`real    0m0.004s`
`user    0m0.004s`
`sys     0m0.000s`

`# time dd bs=8 skip=1 if=1 of=1.tmp oflag=append conv=notrunc`
`real    0m0.590s`
`user    0m0.160s`
`sys     0m0.430s`

mans0n replied:
Wow, that's a way better. Great!

mans0n commented on target/linux/ath79/image/nand.mk:
```diff
@@ -77,6 +90,43 @@ define Device/domywifi_dw33d
+  DEVICE_ALT0_VENDOR := T&Ie
+  DEVICE_ALT0_MODEL := DW02-412H
+  DEVICE_ALT1_VENDOR := KT
+  DEVICE_ALT1_MODEL := GiGA WiFi home
```
The DEVICE_ALT* strings will appear on menuconfig and it will confuse users if the specific model name is not shown there. So I would add `DEVICE_ALT1_VARIANT := (DW02-412H)` here.

And...why did you add `DEVICE_ALT0_VENDOR := T&Ie` here?

rapid-renard replied:
I did it by referring to an [example of this device](https://github.com/openwrt/openwrt/blob/c58005a9431b16107d8c6398b9095a5b3bcfcb25/target/linux/ath79/image/generic.mk#L220) which included both old and new company name...
I have a device that was made after the company name was changed to T&Ie.
However, I am considering deleting this vendor because I think it will make the situation very confusing by adding this.

rapid-renard replied:
> The DEVICE_ALT* strings will appear on menuconfig and it will confuse users if the specific model name is not shown there. So I would add `DEVICE_ALT1_VARIANT := (DW02-412H)` here.

![variant added](https://user-images.githubusercontent.com/79572354/109534351-9a10b480-7afe-11eb-9108-97c5b53d67e3.PNG)
![variant removed](https://user-images.githubusercontent.com/79572354/109534381-a1d05900-7afe-11eb-9807-28b33bab8362.PNG)

I think it's a good suggestion, but this seems to make the target profile too long.

mans0n replied:
> I did it by referring to an example of this device which included both old and new company name...

I would do that only if there's a device with a label "T&Ie DW02-412H".

> I think it's a good suggestion, but this seems to make the target profile too long.

Oh, I didn't know that the original name is appended there... My bad.
BTW menuconfig is notorious for its small menu window... nconfig is better in that perspective.
That name will also appear at the [firmware-selector](https://firmware-selector.openwrt.org/), so I still think we should put DEVICE_ALTx_VARIANT here.

rapid-renard replied:
Okay. I'll remove `T&Ie` vendor and add `DEVICE_ALT0_VARIANT := (DW02-412H)`.

mans0n replied:
Oh, and don't forget to append "{64,128}M" to it. (Wow, it really gets longer and longer...)

mans0n commented on target/linux/ath79/image/nand.mk:
```diff
@@ -77,6 +90,43 @@ define Device/domywifi_dw33d
+  DEVICE_ALT0_MODEL := DW02-412H
+  DEVICE_ALT1_VENDOR := KT
+  DEVICE_ALT1_MODEL := GiGA WiFi home
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-storage kmod-ath10k-ct \
```
I would remove kmod-usb-storage here as it is strictly not a core function of the device. We can install it manually later.

mans0n commented on target/linux/ath79/image/nand.mk:
```diff
@@ -77,6 +90,43 @@ define Device/domywifi_dw33d
+  IMAGE/factory.img := append-kernel | pad-to $$$$(KERNEL_SIZE) | append-ubi | \
+	check-size
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata | \
+	check-size
```
`check-size` compares the image size with `IMAGE_SIZE`, and for sysupgrade.bin comparing a compressed image size with an (uncompressed) flash partition size is not really fair. You can just skip it.

mans0n reviewed (commented):
This will clean up the most things I can find, I think I can provide Reviewed-by after the next round. (I need to take a closer look at the ethernet part)

adschm commented on package/boot/uboot-envtools/files/ath79:
```diff
@@ -74,6 +74,8 @@ domywifi,dw33d)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x8000" "0x10000"
 	;;
+dongwon,dw02-412h-64m|\
+dongwon,dw02-412h-128m|\
```
Alphabetic sorting ...

adschm commented on target/linux/ath79/dts/qca9557_dongwon_dw02-412h-128m.dts:
```diff
@@ -0,0 +1,43 @@
+				label = "ubi";
+				reg = <0x800000 0x6800000>;
+			};
+		};
```
Does this nesting of firmware and kernel/ubi work? Why not just use kernel/ubi?

rapid-renard replied:
It works well. I think this would be good for dumping all areas used by openwrt in NAND, as in the example of NBG6716 ([#](https://github.com/openwrt/openwrt/blob/20b6e014a61a9b40c30fab8184a9abd17f55085b/target/linux/ath79/dts/qca9558_zyxel_nbg6716.dts#L143)).

adschm replied:
I still don't think it's a good idea to have firmware and ubi partition. This whole nesting seems unnecessary to me.

rapid-renard replied:
Okay, I'll revert to original state for this change.

adschm commented on target/linux/ath79/image/nand.mk:
```diff
@@ -77,6 +90,41 @@ define Device/domywifi_dw33d
+define Device/dongwon_dw02-412h-64m
+  $(Device/dongwon_dw02-412h)
+  DEVICE_VARIANT := (64M)
+  DEVICE_ALT0_VARIANT := (DW02-412H) (64M)
```
Why put the name in brackets another time?

rapid-renard replied:
It was a suggestion by @mans0n ([link](https://github.com/openwrt/openwrt/pull/3930#discussion_r585487803)). If you think this isn't right, I'll remove it.

mans0n replied:
I thought what Adrian meant was something like `DEVICE_ALT0_VARIANT := (DW02-412H 64M)`...

mans0n replied:
To be clear,
1. There are three "GiGA WiFi home" devices, DW02-412H, KM06-506H, KM06-704H. Users might be confused if the submodel is not specified.
2. The autocomplete search box of [firmware-selector](https://firmware-selector.openwrt.org/) does not append original name after alt name, so 64/128M images would not be distinguishable at a glance.

adschm replied:
What I meant was I don't see why we need "(DW02-412H)" in variant another time ...

mans0n replied:
Sorry, I misunderstood your words, but don't you think that alt name should be some kind of unique identifier by itself? Or do you think it is just for additional info appended to the original name?

adschm commented on target/linux/ath79/dts/qca9557_dongwon_dw02-412h.dtsi:
```diff
@@ -0,0 +1,165 @@
+
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&art 0x0>;
+	mtd-mac-address-increment = <3>;
```
No address in 0x1002?

rapid-renard replied:
Yes, the values in that location are the same for all devices.

adschm commented on target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -13,6 +13,11 @@ case "$FIRMWARE" in
+	dongwon,dw02-412h-64m|\
+	dongwon,dw02-412h-128m)
+		caldata_extract "art" 0x5000 0x844
+		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary art 0x0) +4)
```
no address in 0x5006?

rapid-renard replied:
Yes, the values in that location are the same for all devices.

mans0n commented on target/linux/ath79/dts/qca9557_dongwon_dw02-412h.dtsi:
```diff
@@ -0,0 +1,165 @@
+&spi {
+	status = "okay";
+
+	num-cs = <1>;
```
We don't need this line. 72bd92bea0aa

mans0n commented on target/linux/ath79/dts/qca9557_dongwon_dw02-412h.dtsi:
```diff
@@ -0,0 +1,165 @@
+			0x54 0x00000000 /* LED Control Register 1 */
+			0x58 0x00000000 /* LED Control Register 2 */
+			0x5c 0x0030c300 /* LED Control Register 3 */
+			0x7c 0x0000007e	 /* PORT0_STATUS */
```
Please reduce two spaces to one.

mans0n commented on target/linux/ath79/dts/qca9557_dongwon_dw02-412h.dtsi:
```diff
@@ -0,0 +1,165 @@
+	status = "okay";
+
+	mtd-mac-address = <&art 0x0>;
+	mtd-mac-address-increment = <1>;
```
From what I remember,
```
   lan: label +2
   wan: label +0/+1
wlan5g: label +4
wlan2g: label +3
```
Could you check the LAN address again?

rapid-renard replied:
label: 9D

lan: 9E (+1)
wan: 9D (+0)
wlan5g: A1 (+4)
wlan2g: A0 (+3)

I set IP allocation policy in stock firmware to router mode, not KT mode.

mans0n replied:
Hmm okay, I haven't checked the mode on the latest stock FW so probably you are right.

adschm replied:
If you checked, please document that in the commit message. While at it, please wrap the commit message to a maximum of 75 characters.

rapid-renard replied:
I didn't realize that I had to keep this... Thanks for informing me about this!

mans0n reviewed (commented):
I'm really sorry for the delay. Now I got some time to review and test this.

Have you also tried kernel v5.10? You'd better add a flash patch for 5.10 as well.

adschm commented on target/linux/ath79/image/nand.mk:
```diff
@@ -77,6 +90,39 @@ define Device/domywifi_dw33d
+
+define Device/dongwon_dw02-412h-64m
+  $(Device/dongwon_dw02-412h)
+  DEVICE_VARIANT := (64M)
```
So, now add `DEVICE_ALT0_VARIANT := (64M)` here and it should be fine (same with 128M of course).

mans0n commented:
Umm...it seems sth got wrong in your patches. (see the title line)
https://github.com/openwrt/openwrt/commit/fe217a892fef0a0f9f5fc44b43589a631532ec1a.patch

mans0n commented on target/linux/ath79/image/nand.mk:
```diff
@@ -1,3 +1,16 @@
+define Build/dw-headers
+	head -c 4 $@ >> $@.tmp && \
+	printf '\x00\x00\x00\x00\x00\x02\x04\x00' >> $@.tmp && \
```
The header code has changed a lot from manatail's original script, but still this line remains almost the same.
If I were you, I would give credit to the original author in the commit message...

Personally I think `head -c 8 /dev/zero >> ...` would suffice here, though.

mans0n replied:
It's not a magic number. It's an internal version code. It will be evident if you compare the value from your nand image dump with the version info from the router setting page. I guess any value will fit there.

rapid-renard replied:
I followed the method you suggested and it really worked. Your assumption was really right. You're such an expert!
I'll update the header generation function as you provided.

mans0n commented:
`Reviewed-by: Sungbo Eo <mans0n@gorani.run>`
`Tested-by: Sungbo Eo <mans0n@gorani.run>`

Thanks for your hard work and patience!

adschm commented on target/linux/generic/pending-5.10/484-mtd-spi-nor-add-esmt-f25l16pa.patch:
```diff
@@ -0,0 +1,11 @@
+--- a/drivers/mtd/spi-nor/esmt.c
++++ b/drivers/mtd/spi-nor/esmt.c
```
If these patches are "pending", please add the full patch as submitted to the kernel list.

adschm commented:
Support patch looks fine. The SPI-NOR support should be sent upstream. If you have received positive feedback there, just update the patches here with the versions sent upstream.

mans0n commented:
@rapid-renard Do you plan to send the spi-nor patch upstream? I can take over if you like (I have my own [version](https://github.com/mans0n/openwrt/commit/6003d0d017b1c811aa79c5a48d2b330827b5cc44#diff-59fa04e73f13fdf73456f053d5afe115b6a6de9b2b7fafc42bbb11d3cb69e796) of it), but it's a rather simple patch and I don't want to take away a good opportunity for getting involved in upstreaming process.

rapid-renard commented:
@mans0n I'm so sorry for the late response. I just finished writing the patch, but I am still holding off on sending it because there was a revision of the model name [F25L16PA -> F25L16PA(2S)] as in [the patch you recently submitted](https://patchwork.ozlabs.org/project/linux-mtd/patch/20210620085639.143694-1-mans0n@gorani.run/).

mans0n commented:
I think you don't need to wait, as the change in flash name is not really critical, and you can always revise or update the patch later if it gets any feedback.

mans0n commented on target/linux/generic/pending-5.10/484-mtd-spi-nor-add-esmt-f25l16pa.patch:
```diff
@@ -0,0 +1,11 @@
+ 
+ static const struct flash_info esmt_parts[] = {
+ 	/* ESMT */
++	{ "f25l16pa", INFO(0x8c2115, 0, 64 * 1024, 64,
```
IIRC the number of sectors should be 32(I mean `64 * 1024, 32`), I've just noticed it.

rapid-renard replied:
Oh, my mistake. I haven't noticed it until now... Thank you for letting me know about it!

mans0n commented:
It seems "-2s" suffix is preferred [there](https://patchwork.ozlabs.org/project/linux-mtd/patch/20210620085639.143694-1-mans0n@gorani.run/#2721911).
Please change the flash name in this PR so that we can go forward.

rapid-renard commented:
I'm sorry for being late... I've modified the previous patch to fit the recent changes and also modified the suffix.

rapid-renard committed:
kernel: add support for ESMT F25L16PA(2S) SPI-NOR

This fixes support for Dongwon T&I DW02-412H which uses F25L16PA(2S) flash.

Signed-off-by: Jihoon Han <rapid_renard@renard.ga>

rapid-renard committed:
ath79: add support for Dongwon T&I DW02-412H

Dongwon T&I DW02-412H is a 2.4/5GHz band 11ac (WiFi-5) router, based on
Qualcomm Atheros QCA9557.

Specifications
--------------

- SoC: Qualcomm Atheros QCA9557-AT4A
- RAM: DDR2 128MB
- Flash: SPI NOR 2MB (Winbond W25Q16DVSSIG / ESMT F25L16PA(2S)) +
         NAND 64/128MB
- WiFi:
  - 2.4GHz: QCA9557 WMAC
  - 5GHz: QCA9882-BR4A
- Ethernet: 5x 10/100/1000Mbps
  - Switch: QCA8337N-AL3C
- USB: 1x USB 2.0
- UART:
  - JP2: 3.3V, TX, RX, GND (3.3V is the square pad) / 115200 8N1

Installation
--------------

1.  Connect a serial interface to UART header and
    interrupt the autostart of kernel.
2.  Transfer the factory image via TFTP and write it to the NAND flash.
3.  Update U-Boot environment variable.
    > tftpboot 0x81000000 <your image>-factory.img
    > nand erase 0x1000000
    > nand write 0x81000000 0x1000000 ${filesize}
    > setenv bootpart 2
    > saveenv

Revert to stock firmware
--------------

1.  Revert to stock U-Boot environment variable.
    > setenv bootpart 1
    > saveenv

MAC addresses as verified by OEM firmware
--------------

   WAN: *:XX (label)
   LAN: *:XX + 1
  2.4G: *:XX + 3
    5G: *:XX + 4

The label MAC address was found in art 0x0.

Credits
--------------

Credit goes to the @manatails who first developed how to port OpenWRT
to this device and had a significant impact on this patch.

And thanks to @adschm and @mans0n for guiding me to revise the code
in many ways.

Signed-off-by: Jihoon Han <rapid_renard@renard.ga>

adschm commented on target/linux/ath79/image/nand.mk:
```diff
@@ -1,3 +1,16 @@
+define Build/dw-headers
+	head -c 4 $@ >> $@.tmp && \
```
For the first command, probably ">" would be better than ">>"?

mans0n replied:
Probably right, I'll prepare a tidy-up patch with that change.

adschm commented:
This has seen enough suffering, I will merge it now. Thanks!

If my recent question is answered yes, we might still change it separately or just leave it.
